insecure/integration/tests/gossipsub/scoring/scoring_test.go

```diff
@@ -0,0 +1,154 @@
+			},
+		},
+		{
+			name: "self-origin, no signature", // bounce back our own messages
+			spamMsgFactory: func(_ peer.ID, victimId peer.ID, topic channels.Topic) *pubsub_pb.Message {
+				return p2ptest.PubsubMessageFixture(t, p2ptest.WithFrom(victimId), p2ptest.WithTopic(topic.String()), p2ptest.WithoutSignature())
+			},
+		},
+		{
+			name: "no sender",
+			spamMsgFactory: func(_ peer.ID, victimId peer.ID, topic channels.Topic) *pubsub_pb.Message {
+				return p2ptest.PubsubMessageFixture(t, p2ptest.WithoutSignerId(), p2ptest.WithTopic(topic.String()))
+			},
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			testGossipSubInvalidMessageDeliveryScoring(t, tc.spamMsgFactory)
+		})
+	}
+}
+
+// testGossipSubInvalidMessageDeliveryScoring tests that when a victim peer is spammed with invalid messages from
+// a spammer peer, the victim will eventually penalize the spammer and stop receiving messages from them.
+// Args:
+// - t: the test instance.
+// - spamMsgFactory: a function that creates unique invalid messages to spam the victim with.
+func testGossipSubInvalidMessageDeliveryScoring(t *testing.T, spamMsgFactory func(peer.ID, peer.ID, channels.Topic) *pubsub_pb.Message) {
+	role := flow.RoleConsensus
+	sporkId := unittest.IdentifierFixture()
+	blockTopic := channels.TopicFromChannel(channels.PushBlocks, sporkId)
+
+	idProvider := mock.NewIdentityProvider(t)
+	spammer := corruptlibp2p.NewGossipSubRouterSpammer(t, sporkId, role, idProvider)
+	ctx, cancel := context.WithCancel(context.Background())
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+	victimNode, victimIdentity := p2ptest.NodeFixture(
+		t,
+		sporkId,
+		t.Name(),
+		idProvider,
+		p2ptest.WithRole(role),
+		p2ptest.WithPeerScoreTracerInterval(1*time.Second),
+		p2ptest.WithPeerScoringEnabled(idProvider),
+	)
+
+	idProvider.On("ByPeerID", victimNode.Host().ID()).Return(&victimIdentity, true).Maybe()
+	idProvider.On("ByPeerID", spammer.SpammerNode.Host().ID()).Return(&spammer.SpammerId, true).Maybe()
+	ids := flow.IdentityList{&spammer.SpammerId, &victimIdentity}
+	nodes := []p2p.LibP2PNode{spammer.SpammerNode, victimNode}
+
+	p2ptest.StartNodes(t, signalerCtx, nodes, 100*time.Millisecond)
+	defer p2ptest.StopNodes(t, nodes, cancel, 2*time.Second)
+
+	p2ptest.LetNodesDiscoverEachOther(t, ctx, nodes, ids)
+	p2ptest.TryConnectionAndEnsureConnected(t, ctx, nodes)
+
+	// checks end-to-end message delivery works on GossipSub
+	p2ptest.EnsurePubsubMessageExchange(t, ctx, nodes, func() (interface{}, channels.Topic) {
+		return unittest.ProposalFixture(), blockTopic
+	})
+
+	totalSpamMessages := 20
+	for i := 0; i <= totalSpamMessages; i++ {
+		spammer.SpamControlMessage(t, victimNode,
+			spammer.GenerateCtlMessages(1),
+			spamMsgFactory(spammer.SpammerNode.Host().ID(), victimNode.Host().ID(), blockTopic))
+	}
+
+	// wait for 3 heartbeats to ensure the score is updated.
```

Comment:
can you add an explanation here why it's necessary to wait for 3 heartbeats to propagate the scoring?

Reply:
https://github.com/onflow/flow-go/pull/4492/commits/56711431ff8e68e0ff90227cefbe26e7c3e6d57b